[Workers] Document wrangler's new unstable_dev API (#5739)
cloudflare/cloudflare-docs: content/workers/wrangler/api.md

---
pcx_content_type: how-to
title: API
weight: 12
---

# Wrangler API

Wrangler offers an experimental API to programmatically manage your Cloudflare Workers.

- [`unstable_dev`](#unstable_dev) - Start a local server for running integration tests against your Worker.

## unstable_dev

Start a local HTTP server for testing your Worker. 

> Once called, `unstable_dev` will return a `fetch` function for invoking your Worker without needing to know the address or port, as well as a `stop` function to shut down the HTTP server.

Author: ```suggestion
Once called, `unstable_dev` will return a `fetch()` function for invoking your Worker without needing to know the address or port, as well as a `stop()` function to shut down the HTTP server.
```

{{<Aside type="note">}}

> The `unstable_dev` function has an `unstable_` prefix because the API may change in the future. 

Author: ```suggestion
The `unstable_dev()` function has an `unstable_` prefix because the API may change in the future. 
```

There are no known bugs at the moment and it is safe to use. If you discover any bugs, please open a [GitHub Issue](https://github.com/cloudflare/wrangler2/issues/new/choose) and we will review the issue.

{{</Aside>}}

### Constructor

```js
const worker = await unstable_dev(script, options, apiOptions)
```

### Parameters

{{<definitions>}}

*   `script` {{<type>}}string{{</type>}}

>     *   A string containing a path to your Worker script.

Reviewer: A path relative to what? I think the current working directory?

Author: Relative to your package.json, I think - not sure how to express that idea

>     *   A string containing a path to your Worker script.

Author: ```suggestion
    *   A string containing a path to your Worker script, relative to your Worker project's root directory.
```

*   `options` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

    *   Optional options object containing `wrangler dev` configuration settings.

*   `apiOptions` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

    *   Optional API options object containing `disableExperimentalWarning`. Set `disableExperimentalWarning` to true to disable Wrangler's warning about using `unstable_` prefixed APIs.

{{</definitions>}}

### Return Type

> `unstable_dev` returns an object containing the following functions:

Reviewer: Should functions that are contained in an object be referred to as methods?

{{<definitions>}}

> *   `fetch` {{<type>}}Promise\<Response>{{</type>}}

Reviewer: Generally I believe that when referring to functions we should include parentheses.

```suggestion
*   `fetch()` {{<type>}}Promise\<Response>{{</type>}}
```

Similarly throughout.

Reviewer: Moreover, could we describe what parameters this function accepts? Especially since we are saying that it is different to a normal `fetch()` (e.g. doesn't need to know the URL and port).

    *   Send a request to your Worker. Returns a Promise that resolves with a [`Response`](/workers/runtime-apis/response) object.

*   `stop()` {{<type>}}Promise\<void>{{</type>}}

    *   Shuts down the dev server. 

{{</definitions>}}

### Usage

> When initiating each test suite, use a `beforeAll` function to start `unstable_dev`. The `beforeAll` function is used to minimize overhead: starting the dev server takes a few hundred milliseconds, starting and stopping for each individual test adds up quickly, slowing your tests down.

Author: ```suggestion
When initiating each test suite, use a `beforeAll()` function to start `unstable_dev`. The `beforeAll` function is used to minimize overhead: starting the dev server takes a few hundred milliseconds, starting and stopping for each individual test adds up quickly, slowing your tests down.
```

> Typically, you'll want to start each test suite with a `beforeAll` function that starts `unstable_dev`. We use the `beforeAll` function because starting the dev server takes a few hundred milliseconds, starting and stopping for each individual test adds up quickly, slowing your tests down.

Reviewer 2: ```suggestion
When initiating each test suite, a `beforeAll` function starts `unstable_dev`. The `beforeAll` function is used to address latency. Starting the dev server takes a few hundred milliseconds, starting and stopping for each individual test adds up quickly, slowing your tests down.
```

Author: ```suggestion
When initiating each test suite, a `beforeAll` function starts `unstable_dev`. The `beforeAll` function is used to minimize overhead: starting the dev server takes a few hundred milliseconds, starting and stopping for each individual test adds up quickly, slowing your tests down.
```
Maybe more like this @deadlypants1973 ?

Reviewer: How about?

```suggestion
When initiating each test suite, use a `beforeAll` function to start `unstable_dev`. The `beforeAll` function is used to minimize overhead: starting the dev server takes a few hundred milliseconds, starting and stopping for each individual test adds up quickly, slowing your tests down.
```

> In each test case, call `await worker.fetch`, and check that the response is what you expect. 

Reviewer: ```suggestion
In each test case, call `await worker.fetch()`, and check that the response is what you expect. 
```

> In each test case you'll call `await worker.fetch`, and check that the response is what you expect. 

Reviewer 2: ```suggestion
In each test case, call `await worker.fetch`, and check that the response is what you expect. 
```

> To wrap up a test suite, call `await worker.stop` in an `afterAll` function.

Reviewer: ```suggestion
To wrap up a test suite, call `await worker.stop()` in an `afterAll` function.
```

(you get the idea)

> To wrap up a test suite, you should call `await worker.stop` in an `afterAll` function.

Reviewer 2: ```suggestion
To wrap up a test suite, call `await worker.stop` in an `afterAll` function.
```

{{<tabs labels="js | ts">}}

> {{<tab label="js" default="true">}}

Reviewer: Should we not be encouraging people to use TS by making it the default?

Author: Apparently not, no @petebacondarwin:
<img width="738" alt="image" src="https://user-images.githubusercontent.com/3822106/189870351-fbb57e17-71b0-4b3a-b2a1-6dd8467d3ed2.png">

```js
---
filename: src/index.test.js
---

> import { unstable_dev } from 'wrangler'

Author: ```suggestion
const { unstable_dev } = require("wrangler");
```

---
pcx_content_type: how-to
title: API
weight: 12
---

# Wrangler API

> Wrangler offers an API to programmatically manage your Cloudflare Workers

Reviewer: ```suggestion
Wrangler offers an experimental API to programmatically manage your Cloudflare Workers
```

- [`unstable_dev`](#unstable_dev) - Start a local server for running integration tests against your Worker.

## unstable_dev

> Start a local server for testing your Worker.

Reviewer: Could you expand a bit to explain that this literally starts an HTTP server but that you can use the `worker.fetch()` method to interact with the server without having to know about the ports, etc.

{{<Aside type="note">}}

The `unstable_dev` function has an `unstable_` prefix because the API may change in the future. There are no known bugs at the moment and it is safe to use. If you discover any bugs, please open a [GitHub Issue](https://github.com/cloudflare/wrangler2/issues/new/choose) and we'll fix them as soon as possible.

{{</Aside>}}

### Constructor

```js
const worker = await unstable_dev(script, options, apiOptions)
```

### Parameters

{{<definitions>}}

*   `script` {{<type>}}string{{</type>}}

    *   A string containing a path to your Worker script.

*   `options` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

    *   Optional options object containing `wrangler dev` configuration settings.

*   `apiOptions` {{<type>}}object{{</type>}} {{<prop-meta>}}optional{{</prop-meta>}}

>     *   Optional API options object containing `disableExperimentalWarning`. Set `disableExperimentalWarning` to true to disable wrangler's warning about using `unstable_` prefixed APIs.

Reviewer: And `testMode`?

Author: Intentionally not documenting `testMode`, tbh it's there because we let Pages use unstable_dev before we had time to finish building it for tests

Reviewer: 👍

> 

Reviewer: Can we have a section describing the return type and its methods?

{{</definitions>}}

> ### Usage

Reviewer: Can we have a little preamble to the usage, that explains about how and when you need to call `worker.stop()`
